### PR TITLE
unifying device_mesh init to enable PP + TP inference

### PR DESCRIPTION
**Pull request by 3outeille**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48155) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48155)

# Scripts

```python
"""Sanity check: run a model with transformers using PP + TP.

Usage:
    torchrun --nproc-per-node=4 pp_tp_inference.py
"""
import os

import torch

from transformers import AutoModelForCausalLM, AutoTokenizer
from transformers.distributed import DistributedConfig


def main():

    rank = int(os.environ.get("RANK", 0))
    world_size = int(os.environ.get("WORLD_SIZE", 1))
    # Qwen2.5-0.5B has 14 attention heads and 2 KV heads, so TP must divide both.
    tp_size = 2
    pp_size = 2

    # model_id = "TinyLlama/TinyLlama-1.1B-Chat-v1.0"
    model_id = "Qwen/Qwen2.5-0.5B-Instruct"
    is_tie_word_embeddings = model_id == "Qwen/Qwen2.5-0.5B-Instruct"

    print(f"Rank {rank} of {world_size} using {pp_size} pipeline stages and {tp_size} tensor parallel stages")
    tokenizer = AutoTokenizer.from_pretrained(model_id)
    model = AutoModelForCausalLM.from_pretrained(
        model_id,
        distributed_config=DistributedConfig(pp_size=pp_size, tp_size=tp_size),
        dtype=torch.bfloat16,
        tie_word_embeddings=is_tie_word_embeddings,
    ).eval()

    prompts = [
        "The capital of France is",
        "In one sentence, explain what pipeline parallelism is:",
    ]

    for prompt in prompts:
        input_ids = tokenizer(prompt, return_tensors="pt").input_ids.to(model.device)
        output_ids = model.generate(input_ids, max_new_tokens=64, do_sample=False)
        if rank == 0:
            print(f"PROMPT : {prompt!r}")
            print(f"OUTPUT : {tokenizer.decode(output_ids[0], skip_special_tokens=True)!r}")
            print("-" * 80)


if __name__ == "__main__":
    main()
```

```
(env_2D_pp_tp) ➜  transformer_2D_pp_tp git:(transformer_2D_pp_tp) ✗ torchrun --nproc-per-node=4 pp_tp_inference.py
W0911 10:51:40.139000 2572005 torch/distributed/run.py:874] 
W0911 10:51:40.139000 2572005 torch/distributed/run.py:874] *****************************************
W0911 10:51:40.139000 2572005 torch/distributed/run.py:874] Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed. 
W0911 10:51:40.139000 2572005 torch/distributed/run.py:874] *****************************************
Rank 3 of 4 using 2 pipeline stages and 2 tensor parallel stages
Rank 0 of 4 using 2 pipeline stages and 2 tensor parallel stages
Rank 2 of 4 using 2 pipeline stages and 2 tensor parallel stages
Rank 1 of 4 using 2 pipeline stages and 2 tensor parallel stages
Loading weights: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 145/145 [00:00<00:00, 2649.77it/s]
Loading weights: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 145/145 [00:00<00:00, 2795.83it/s]
Loading weights: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 146/146 [00:00<00:00, 2603.98it/s]
Loading weights: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 146/146 [00:00<00:00, 2815.66it/s]
PROMPT : 'The capital of France is'
OUTPUT : 'The capital of France is Paris. It was founded in 789 AD by Charlemagne, the last king of the Carolingian dynasty. The city has a long and rich history dating back to the Roman Empire. In fact, itwas the first major European city to be built on land that would later become Paris.\nParis is'
--------------------------------------------------------------------------------
PROMPT : 'In one sentence, explain what pipeline parallelism is:'
OUTPUT : 'In one sentence, explain what pipeline parallelism is: Pipeline parallelism refers to the ability of a system to execute multiple tasks concurrently within a single pipeline. It involves using multiple processors or threads to process data simultaneously, allowing for faster processing times and improved efficiency in certain applications. This technique can be particularly useful in scenarios where data needs to be processed quickly, such as in real'
--------------------------------------------------------------------------------
```

# Benchmark on 1 node, `device_map=auto` vs `PP=2 & TP=4` 

The main advantage is that PP + TP can scale to > 1 node while device_map=auto is only for 1 node

image
image
image

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
LGTM I need to see it in practice this, an example script or a .md that shows how to?

**Comment by 3outeille** [comment]
ill add in the PR how to use it

File: src/transformers/distributed/pipeline_parallel.py
Comment (molbap) [inline src/transformers/distributed/pipeline_parallel.py]: on `ForConditionalGeneration` models, base models do not have an attribute layers, will not work ootb for non-decoder-only models right?

File: src/transformers/distributed/pipeline_parallel.py [hunk 1]
```diff
@@ -32,6 +32,24 @@
     import torch.distributed as dist
 
 
+# Duplicated from generation.utils to avoid a circular import.
+MULTIMODAL_INPUTS_TO_DROP_OUTSIDE_PREFILL = (
```
Comment (3outeille) [inline src/transformers/distributed/pipeline_parallel.py:36]: mhmm is there another way ? if a variable change in `generation.utils`, change wont be reflected here

**Comment by 3outeille** [comment]
will revert the VLM commit as I need this PR for TRL. It needs a bit more work to make it work for every VLMs out there !

File: src/transformers/distributed/utils.py [hunk 2]
```diff
@@ -129,98 +129,37 @@ def _distributed_barrier():
...
+
+    if not mesh_shape:
+        return None, None
 
     device_type = torch._C._get_accelerator().type
-    _ensure_torch_distributed(device_type)
+    if distributed_config.tp_size > 1 and device_type == "mps":
+        raise RuntimeError("Tensor parallelism is not supported on MPS devices.")
 
+    _ensure_torch_distributed(device_type)
```
Comment (ArthurZucker) [inline src/transformers/distributed/utils.py:156]: Nice unification! 
I have not tested but the simpler the better!

File: src/transformers/distributed/configuration_utils.py [hunk 3]
```diff
@@ -73,11 +73,10 @@ def __post_init__(self):
...
             self.tp_size = 1
 
-        if self.tp_size > 1 and self.fsdp_size > 1 and self.pp_size > 1:
+        if self.fsdp_size > 1 and (self.tp_size > 1 or self.pp_size > 1):
             raise ValueError(
-                "FSDP+TP+PP is not supported yet. "
-                "Use DistributedConfig(fsdp_size=N) or DistributedConfig(tp_size=N) or DistributedConfig(pp_size=N), not all three. "
-                "Only 1D support is available for now."
+                "Combining FSDP with tensor or pipeline parallelism is not supported yet. "
+                "Use FSDP alone, or combine tensor and pipeline parallelism."
```
Comment (ArthurZucker) [inline src/transformers/distributed/configuration_utils.py]:
```suggestion
                "Combining FSDP with tensor or pipeline parallelism is not supported yet. "
                "Use FSDP alone, or combine TP and PP."
```

**Review by ArthurZucker — APPROVED** [review]
Ty I like that it simplifies
